app/models/grda_warehouse/cohort_column_type.rb

```diff
@@ -0,0 +1,310 @@
+# frozen_string_literal: true
+
+module GrdaWarehouse
+  class CohortColumnType < GrdaWarehouseBase
+    validates :class_name, presence: true, uniqueness: true
+
+    scope :active, -> { where(active: true) }
+
+    def activate
+      update(active: true)
+    end
+
+    def deactivate
+      update(active: false)
+      remove_from_cohorts
+    end
+
+    def remove_from_cohorts
+      GrdaWarehouse::Cohort.all.each do |cohort|
+        cohort.update!(column_state: cohort.column_state.reject { |col| col.cohort_column_type.class_name == class_name })
+      end
+    end
+
+    def self.maintain!
+      known_cohort_column_types.each do |class_name|
+        column_type = find_by(class_name: class_name)
+        create!(class_name: class_name, active: true) unless column_type
```

Comment:
Maybe find_or_create here?  I think it should work, though I kind of like your implementation better, since you are just throwing out the record if found. Maybe that should be:

```suggestion
        create!(class_name: class_name, active: true) unless where(class_name: class_name).exists?
```
To prevent instantiating the object. 

Actually, on second thought, I think you want to pluck all existing class names and to the check against the result (converted to a set). Then this will only make one query when no new columns are needed.

Reply:
updated [here](https://github.com/greenriver/hmis-warehouse/pull/5299/commits/e36fbed7fbb24bc00231cb8cfca030ef129a67e7) to pluck the existing names ahead of time.